utilsForFrontend.mjs

```diff
@@ -458,28 +458,44 @@ export function sigmoid(x) {
 }
 
 /**
- * Return whether an element is practically visible, considing things like 0
- * size or opacity, ``display: none``, and ``visibility: hidden``.
+ * Return whether an element is practically visible, considering things like 0
+ * size or opacity, ``visibility: hidden`` and ``overflow: hidden``.
+ *
+ * This could be 5x more efficient if https://github.com/w3c/csswg-drafts/issues/4122
+ * happens.
  */
 export function isVisible(fnodeOrElement) {
     const element = toDomElement(fnodeOrElement);
+    // Avoid reading ``display: none`` due to Bug 1381071
+    const elementRect = element.getBoundingClientRect();
+    if (elementRect.width === 0 && elementRect.height === 0) {
+        return false;
+    }
+    const elementStyle = getComputedStyle(element);
+    if (elementStyle.visibility === 'hidden') {
+        return false;
+    }
+    const frame = element.ownerDocument.defaultView;
     for (const ancestor of ancestors(element)) {
-        const style = getComputedStyle(ancestor);
-        if (style.visibility === 'hidden' ||
-            style.display === 'none' ||
-            style.opacity === '0' ||
-            style.width === '0' ||
-            style.height === '0') {
+        const isElement = ancestor === element;
+        const style = isElement ? elementStyle : getComputedStyle(ancestor);
+        if (style.opacity === '0') {
+            return false;
+        }
+        if (style.display === 'contents') {
+            // display: contents elements have no box themselves, but children are
+            // still rendered.
+            continue;
+        }
+        const rect = isElement ? elementRect : ancestor.getBoundingClientRect();
+        if ((rect.width === 0 || rect.height === 0) && elementStyle.overflow === 'hidden') {
+            // Zero-sized ancestors don’t make descendants hidden unless the descendant
+            // has overflow: hidden
+            return false;
+        }
+        // Check if the element is off-screen
+        if (isElement && ((rect.right + frame.scrollX < 0) || (rect.bottom + frame.scrollY < 0))) {
```

Comment:
Unless I'm mistaken, you can move this to before the `for` loop, strip off the `isElement` condition, and it'll be equivalent.

Reply:
Yes, I can do that. I also modified the check, as previously, I hadn't checked all four conditions (element is above, to the left, to the right or below the visible portion of the page).